csharp/src/Google.Protobuf/Reflection/FileDescriptor.cs

```diff
@@ -315,6 +311,49 @@ private void CrossLink()
             }
         }
 
+        /// <summary>
+        /// Converts the given descriptor binary data into FileDescriptor objects.
+        /// Note: reflection using the returned FileDescriptors is not currently supported.
+        /// </summary>
+        /// <param name="descriptorData">The binary file descriptor proto data. Must not be null, and any
+        /// dependencies must come before the descriptor which depends on them. (If A depends on B, and B
+        /// depends on C, then the descriptors must be presented in the order C, B, A.) This is compatible
+        /// with the order in which protoc provides descriptors to plugins.</param>
+        /// <returns>The file descriptors corresponding to <paramref name="descriptorData"/>.</returns>
+        public static IReadOnlyList<FileDescriptor> FromDescriptorData(IEnumerable<ByteString> descriptorData)
```

Comment:
That feels like it's more work for users, certainly in the case where they already have all the file descriptors in one go (which is definitely the case for the compiler plugin use case I'm pursuing at the moment). Additionally, `DescriptorPool` is currently internal - making it public would be a bigger change.

We could potentially make that an *option* later if we wanted, and make use of it from this code as well to avoid creating multiple pools, but it's definitely a bigger change.

At the moment I believe we only populate the `DescriptorPool` on construction of `FileDescriptors`, so we don't need to worry about concurrent access to it etc. Making it externally mutable but still thread-safe would be more of a challenge. Feasible, I'm sure - but I'd like to hear more of a user-driven benefit before diving into the code. (It's been a long time since I wrote this code, and it'll take quite a while to get my head round the aims of it in order to be confidence in what I'm doing.)

Reply:
How much work to make the descriptor pool public? Is it blocked by anything? Make descriptor pool public is the correct way we should go with. 

From protobuf's review, we should try to make the APIs between different languages similar. There's no such API for FileDescriptor in another languages which can build a FileDescriptorProto, and corresponding functionality is in descriptor pool. It is not good to make C# unique from others.

Reply:
It's blocked by "It's another public API, and making anything public is difficult."

Note that other languages are already different, by virtue of `FileDescriptorProto` being public, which it can't be in C# at the moment. Personally I don't think APIs *should* always be the same - there are different idioms in different languages, and they should generally be embraced over uniformity at the cost of simplicity and familiarity for a particular language.

I'll have a look tomorrow at how much of `DescriptorPool` we'd have to expose (and what it looks like in Java), but it does seem odd to force users to learn about a new concept that doesn't help them. (The user doesn't want a descriptor pool. They just want the `FileDescriptor` objects corresponding to the data they've got. The descriptor pool is an implementation detail, surely?)

Reply:
Looking at the Java code, DescriptorPool is private there as well. The two ways I've found of obtaining a `FileDescriptor` in Java are:

```java
public static FileDescriptor buildFrom(
  final FileDescriptorProto proto, final FileDescriptor[] dependencies) { ... }

public static FileDescriptor buildFrom(
  final FileDescriptorProto proto, final FileDescriptor[] dependencies,
  final boolean allowUnknownDependencies) { ...}
```

I could make the C# code match those signatures as far as possible - although as `FileDescriptorProto` isn't exposed, they'd have to be `ByteString`s. I'd *prefer* to also keep the "convert all of these descriptors that are already in order" as well if possible, given that that's the most convenient approach for proto plugins. Let me know what you think.

Reply:
Agree with it. Maybe use a similar name like buildFromByte() or something

Reply:
I'd quite like `FromByteStrings` - we've used "FromX" in multiple places (e.g. `ByteString.FromBase64`) and that tends to be the convention in .NET rather than "BuildFromX". But definitely something indicating that it's bytes would be good as that leaves other names open for when we (hopefully) get to expose `FileDescriptorProto` when we support proto2.

Would you be okay with `FromByteStrings`?

Reply:
In protobuf, we use ParseFromX() when parse data to message and BuildFrom() when build FileDescriptorProto to FileDescriptor. I think this is different for FromX(ByteString.FromBase64): it involves other proto files and need to do cross link between thenm which more like a build.

But I am open to the naming if you still think FromByteStrings() is better for C#

Reply:
I still prefer `FromByteStrings`, but I'm happy enough with `Build` if you want to convey that it's doing more work. (I note that the existing method called by generated code is `FromGeneratedCode` but that's not a problem - we don't expect that to be called manually anyway.)

I'll make the change and push it shortly. Anything else that needs changing before I merge, when I've changed that name?